### ISSUE TITLE
autocorrect on save

### ISSUE THREAD

**Issue by esalter** [misogi/vscode-ruby-rubocop#49]
It seems like this plugin only lints on save, not autocorrect.  I'm used to being able to just save a file and have it auto-run formatters to correct mistakes. Can that be added?

This is somewhat related to https://github.com/misogi/vscode-ruby-rubocop/issues/48  since I wouldn't want to lose undo history in this case though.

Thanks!

**Comment by damien** [comment]
@esalter I'm not sure I understand your request, could you clarify?

When you say "lint on autocorrect", do you mean that Rubocop should re-lint after autocorrecting a file?

**Comment by esalter** [comment]
I mean that the act of saving the file should auto-correct it as well.  It does not appear to do that right now.

**Comment by damien** [comment]
@esalter You are correct. We could plan this as a feature, but it'd probably be opt-in via a config option of some sort.

**Comment by esalter** [comment]
That would be fine.  I took a stab at it locally and it seems easy (just append the option if configured), but I'd want to undo functionality mentioned in the other issue which I'm not sure how to do.  Thanks for taking a look. :)

**Comment by damien** [comment]
@esalter Would you be interested in contributing once #48 was resolved? The more the merrier, as they say!

**Comment by esalter** [comment]
I'd be happy to, just ping me.  Thanks!

**Comment by tjgillies** [comment]
>That would be fine. I took a stab at it locally and it seems easy

@esalter Is there a branch with your changes?

**Comment by gdxn96** [comment]
Any updates on this?

**Comment by ippa** [comment]
autoformat/correct on save would indeed be nice, just like I get with prettier when working with JS.

**Comment by crispinheneise** [comment]
I tried to get this to work using [editor.codeActionsOnSave](https://hk.saowen.com/a/878d75f3829ad44478e9b7d9b510e3367c8e91048384f868f5e8a3c32dd6348f) like this: 

    "editor.codeActionsOnSave": {
        "ruby.rubocopAutocorrect": true
    },

... but it does nothing when saved. I also tried scoping the language, but same result: 

    "[ruby]": {
        "editor.codeActionsOnSave": {
            "ruby.rubocopAutocorrect": true
        }
    },

**Comment by rtoma** [comment]
following is working for me:
```
{
    "editor.formatOnSave": true,
    "editor.formatOnSaveTimeout": 5000,
    "ruby.rubocop.executePath": "path/where/rubocop/is/located",
    "ruby.format": "rubocop",
}
```

**Comment by rashmi-tondare** [comment]
> following is working for me:
> 
> ```
> {
>     "editor.formatOnSave": true,
>     "editor.formatOnSaveTimeout": 5000,
>     "ruby.rubocop.executePath": "path/where/rubocop/is/located",
>     "ruby.format": "rubocop",
> }
> ```

I'm unable to get the autocorrection to work. Tried the settings mentioned above, but vscode throws up an error for `editor.formatOnSaveTimeout` saying `Unknown configuration setting`. Anyone else facing this issue or am I missing something?

**Comment by rtoma** [comment]
> I'm unable to get the autocorrection to work. Tried the settings mentioned above, but vscode throws up an error for `editor.formatOnSaveTimeout` saying `Unknown configuration setting`. Anyone else facing this issue or am I missing something?

To be sure: that piece of config is my in `<projectdir>/.vscode/settings.json` - works for me with vscode 1.28.2

**Comment by rdyson** [comment]
I was able to get this working using @rtoma's settings minus the `ruby.rubocop.executePath` per [this answer](https://stackoverflow.com/questions/53367947/rubocop-on-vscode-not-working-error-rubocop-is-not-executable) on Stack Overflow. If an `executePath` is not specified, PATH will be searched.

Here's what is working for me, VS Code version 1.33.0, ruby-rubocop version 0.7.1.

```
"editor.formatOnSave": true,
"editor.formatOnSaveTimeout": 5000,
"ruby.format": "rubocop",
```

**Comment by DanielTetration** [comment]
@rdyson I tried this and `ruby.format` is an 'unknown configuration setting'

**Comment by rdyson** [comment]
> d

Perhaps that setting isn't required — `ruby.format` is a setting for the [Ruby extension](https://marketplace.visualstudio.com/items?itemName=rebornix.Ruby).

Try removing the config settings I mention above (all 3 lines) and make sure you have the latest version of the ruby-rubocop extension, which is 0.8.0. Rubocop is invoked on save successfully with this setup, with no additional settings in VS Code config. I verified this by starting with a blank settings.json.

Hope that helps!

**Comment by boddhisattva** [comment]
For one's future reference, I was able to do it by adding settings present [here](https://github.com/boddhisattva/dotfiles/blob/master/vscode/settings.json) & have written a little documentation about how to add the settings in VS code [here](https://github.com/boddhisattva/dotfiles/blob/master/vscode/what_each_settings_mean.md)

**Comment by kirrg001** [comment]
I have used in my settings.json

```
    "[ruby]": {
      "editor.formatOnSave": true
    },
```

Works fine, but the auto-correction kicks in very very late. I have to wait a couple of seconds. Any idea why?

**Comment by CapsuleMaglev** [comment]
If anyone has problems with this, I used bundler and it ran fine. Maybe it's some problem with virtual environments like RVM or Rbenv. Cheers!

`"ruby.rubocop.useBundler": true`

**Comment by carlows** [comment]
Same issue, even with `useBundler` it takes seconds to save the file 🤔

**Comment by Benzoh** [comment]
I solved it by writing as follows.

```
  "[ruby]": {
    "editor.defaultFormatter": "misogi.ruby-rubocop"
  }
```

**Comment by ethanneff** [comment]
Can confirm, this worked for me:

- Have `ruby-rubocop` and `ruby` vscode extensions installed
- Add `.vscode/settings.json` at the root.
  ```json 
  {
    "editor.formatOnSave": true,
    "[ruby]": {
      "editor.defaultFormatter": "misogi.ruby-rubocop"
    }
  }
  ```
